tedge/src/cli/certificate/error.rs

```diff
@@ -100,3 +105,76 @@ impl CertError {
+//         },
+//     },
+// )
+// This chain may break if underlying crates change.
+pub(crate) fn get_webpki_error_from_reqwest(err: reqwest::Error) -> CertError {
+    if let Some(rustls::TLSError::WebPKIError(cert_validation_error)) = err
+        // get `hyper::Error::Connect`
+        .source()
+        .and_then(|hyper_error| hyper_error.downcast_ref::<hyper::Error>())
+        .and_then(|hyper_error| hyper_error.source())
+        // Surprise: `Custom` type is `std::io::Error`; this is our first `Custom`.
+        .and_then(|connect_error| connect_error.downcast_ref::<std::io::Error>())
+        // A shortcut to get ref to our error 2 layers down.
+        .and_then(|custom_error| custom_error.get_ref())
+        // This is our second `Custom`.
+        .and_then(|custom_error2| custom_error2.downcast_ref::<std::io::Error>())
+        // Get final error type from `Custom`.
+        .and_then(|custom_error2| custom_error2.get_ref())
+        .and_then(|webpki_error| webpki_error.downcast_ref::<rustls::TLSError>())
+    {
+        match cert_validation_error {
+            webpki::Error::CAUsedAsEndEntity => CertError::WebpkiValidation {
+                hint: "A CA certificate is used as an end-entity server certificate. Make sure that the certificate used is an end-entity certificate signed by CA certificate.".into(),
+                msg: cert_validation_error.to_string(),
+            },
+
+            webpki::Error::CertExpired => CertError::WebpkiValidation {
+                hint: "The server certificate has expired, the time it is being validated for is later than the certificate's `notAfter` time.".into(),
+                msg: cert_validation_error.to_string(),
+            },
+
+            webpki::Error::CertNotValidYet => CertError::WebpkiValidation {
+                hint: "The server certificate is not valid yet, the time it is being validated for is earlier than the certificate's `notBefore` time.".into(),
+                msg: cert_validation_error.to_string(),
+            },
+
+            webpki::Error::EndEntityUsedAsCA => CertError::WebpkiValidation {
+                hint: "An end-entity certificate is used as a server CA certificate. Make sure that the certificate used is signed by a correct CA certificate.".into(),
+                msg: cert_validation_error.to_string(),
+            },
+
+            webpki::Error::InvalidCertValidity => CertError::WebpkiValidation {
+                hint: "The server certificate validity period (`notBefore`, `notAfter`) is invalid, maybe the `notAfter` time is earlier than the `notBefore` time.".into(),
+                msg: cert_validation_error.to_string(),
+            },
+
+            webpki::Error::UnknownIssuer => CertError::WebpkiValidation {
+                hint: r#"The server certificate's issuer is unknown, this may be caused by the server certificate's issuer not being trusted on your system.
+To add a certificate to trusted list follow the guide on:
```

Comment:
`To add a certificate to trusted list` -> `To add a certificate to the trusted list`